### ISSUE TITLE
Wpf: ChromiumWebBrowser throws a NullReferenceException if its ActualSize changes before it appears in the VisualTree

### ISSUE THREAD

**Issue by jamespearce2006** [cefsharp/CefSharp#985]
If a ChromiumWebBrowser is created outside the visual tree (because it's in a tab that's not yet been opened, for example), any changes to its ActualSize throw an exception. This might happen if the tab in question provides a preview tooltip, or needs to render the control into a RenderTargetBitmap for some other reason.

The exception occurs in CreateOffscreenBrowserWhenActualSizeChanged(), and occurs because source is still null.

Some example code that demonstrates the problem.

    private void Button_Click(object sender, RoutedEventArgs e)
    {
        const double width = 640d;
        const double height = 480d;

        ChromiumWebBrowser webBrowser = new ChromiumWebBrowser();
        webBrowser.Address = "www.google.com";

        Viewbox viewBox = new Viewbox();
        viewBox.Child = webBrowser;
        viewBox.Measure(new Size(width, height));
        viewBox.Arrange(new Rect(0, 0, width, height));
        viewBox.UpdateLayout();

        RenderTargetBitmap bitmap = new RenderTargetBitmap((int) width, (int) height, 96, 96, PixelFormats.Pbgra32);
        bitmap.Render(viewBox);

        Image img = new Image();
        img.Source = bitmap;

        content.Content = img;
    }

Additionally, content seems to reload as the ChromiumWebBrowser is added and removed from the VisualTree.

**Comment by rassilon** [comment]
This appears to be about needing to ensure that the Viewbox/RenderTargetBitmap has an associated `HwndSource` somehow I think. Either that, or CefSharp creating the HwndSource with an empty parent for these purposes. If the WPF control moves into the Visual tree afterwards, I would hope that the event handlers would recognize the transition and update `source` appropriately.

I think...

Not even remotely a WPF expert,
Bill

**Comment by jamespearce2006** [comment]
If the element is not in the Visual Tree, it's not possible to get an
HwndSource from it.

It's not clear to me (I'm new to Cef and CefSharp) what this Hwnd is
ultimately used for, but it seems like it would be safer to create a
dedicated, non visible, Hwnd for this purpose...

The other, related, problem is that removing and re-adding a
ChromiumWebBrowser to a visual tree causes it to completely refresh. This
is problematic if you use them in a Tab Control. I see the WpfExample has a
specialist TabControl to specifically circumvent this problem, but that's
not always an option.

James

**Comment by amaitland** [comment]
> It's not clear to me (I'm new to Cef and CefSharp) what this Hwnd is ultimately used for, but it seems like it would be safer to create a dedicated, non visible, Hwnd for this purpose...

It's preferably to pass in the `Hwnd` for the `Window` though not ultimately required by `CEF`, if the source is null passing in `IntPtr.Zero` will work (`OffScreen` browser doesn't require a `Hwnd` see https://github.com/cefsharp/CefSharp/blob/master/CefSharp.OffScreen/ChromiumWebBrowser.cs#L97)

> The other, related, problem is that removing and re-adding a ChromiumWebBrowser to a visual tree causes it to completely refresh. This is problematic if you use them in a Tab Control. I see the WpfExample has a specialist TabControl to specifically circumvent this problem, but that's not always an option.

By all means investigate this further. If you come up with a workable solution then please submit a `PR`

**Comment by rassilon** [comment]
What I meant was, that HwndSources can be created yourself. (since the constructor/type are public)

However, amaitland's point is more useful, if source.xxx is null, we should just pass in IntPtr.Zero I would imagine.

**Comment by amaitland** [comment]
> What I meant was, that HwndSources can be created yourself. (since the constructor/type are public)

Fair point, we could manually create one. Don't think it's required though.

**Comment by rassilon** [comment]
> The other, related, problem is that removing and re-adding a ChromiumWebBrowser to a visual tree causes it to completely refresh. This is problematic if you use them in a Tab Control. I see the WpfExample has a specialist TabControl to specifically circumvent this problem, but that's not always an option.

So, WpfExample appears to work around complete destruction and recreation of child ContentPresenters. (i.e. by caching the ChromiumWebBrowser control instance as a side effect of NonReloadingTabControl)

That isn't the same thing as having a ChromiumWebBrowser move from one PresentationSource to another. (which only appears to readd the SourceHook, and notify CEF of a possible resize/whatever) 

Now, I'm not sure how easy that is to arrange for yourself in WPF. I hope it wouldn't be too terribly bad. 

However, if the calls to CEF's NotifyScreenInfoChanged (via PresentationSourceChangedHandler) or CEF's WasHidden (via `OnIsVisibleChanged`) are called more often than necessary and you think a Freeze/Thaw approach might work that might be something that could be approached.

**Comment by amaitland** [comment]
> Freeze/Thaw approach might work

Sounds like a reasonable approach :+1:

**Comment by jamespearce2006** [comment]
I can confirm that falling back to IntPtr.Zero, if source is null, fixes the NullReferenceException, and everything seems to work.

    private void CreateOffscreenBrowserWhenActualSizeChanged()
    {
        if (browserCreated || System.ComponentModel.DesignerProperties.GetIsInDesignMode(this))
        {
            return;
        }

        managedCefBrowserAdapter.CreateOffscreenBrowser(source == null ? IntPtr.Zero : source.Handle, BrowserSettings, Address);
        browserCreated = true;
    }

Question - is there any particular reason why it can't always be IntPtr.Zero? What's the benefit in using a genuine Hwnd?

It looks like my second problem is entirely self-inflicted...

**Comment by amaitland** [comment]
`The |parent| value will be used to identify monitor info and to act as the parent window for dialogs, context menus, etc. If |parent| is not provided then the main screen monitor will be used and some functionality that requires a parent window may not function correctly. `

http://magpcss.org/ceforum/apidocs3/projects/%28default%29/CefWindowInfo.html#SetAsWindowless%28cef_window_handle_t,bool%29

**Comment by rassilon** [comment]
Hrm. I wonder if you can call `SetAsWindowless` more than once. Probably not.

**Comment by amaitland** [comment]
It's just a helper method on `CefWindowInfo`,

```c++
void SetAsWindowless(CefWindowHandle parent, bool transparent) {
    windowless_rendering_enabled = TRUE;
    parent_window = parent;
    transparent_painting_enabled = transparent;
}
```

**Comment by rassilon** [comment]
I meant, I wonder if anyone will notice if the parent_window changes after you call this method a Nth time.

**Comment by rassilon** [comment]
(Since I might imagine that for offscreen rendering SetAsWindowless might make a difference if you reparent a WPF ChromiumWebBrowser onto a different (just to make life complicated) DPI monitor.)

**Comment by amaitland** [comment]
`CefWindowInfo` is just a simple struct that gets passed into `CefBrowserHost::CreateBrowser()`, so I doubt it would change anything.

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/ManagedCefBrowserAdapter.h#L81

> (Since I might imagine that for offscreen rendering SetAsWindowless might make a difference if you reparent a WPF ChromiumWebBrowser onto a different (just to make life complicated) DPI monitor.)

There's a `NotifyScreenInfoChanged()` method that needs to be called in this scenario. 
https://github.com/cefsharp/CefSharp/pull/937/files#diff-bfec65df46ec49f4ed79472147689599R901

When it's finished #937 should handle that scenario.

**Comment by jamespearce2006** [comment]
Just wondering - what should I do about the proposed fix to the NullReferenceException above (in CreateOffscreenBrowserWhenActualSizeChanged)...

**Comment by amaitland** [comment]
Create a `PR`.

**Comment by jamespearce2006** [comment]
Done - it's my first ever one...

**Comment by amaitland** [comment]
Merged #993

**Comment by amaitland** [comment]
> Done - it's my first ever one..

:+1:
